<img src="https://www.upsieutoc.com/images/2020/08/13/backgroundted.jpg" alt="Parrondo" border="0">

- TED (Technology, Entertainment, Design - Công nghệ, Giải trí, Thiết kế) là một tổ chức truyền thông đại chúng chuyên đăng tải những phần nói chuyện, cho phép xem trực tuyến miễn phí, với khẩu hiệu "ideas worth spreading" ("Những ý tưởng đáng lan truyền"). TED được thành lập vào tháng 2 năm 1984 dưới dạng một buổi hội thảo, và bắt đầu được tổ chức hàng năm từ năm 1990. Ban đầu, nội dung của TED nhấn mạnh vào công nghệ và thiết kế, nhưng nó đã mở rộng trọng tâm của nó bao gồm các cuộc thảo luận về nhiều chủ đề khoa học, văn hoá và học thuật.
- Hội thảo chính của TED được tổ chức thường niên tại Vancouver, British Columbia, Canada ở Trung tâm Hội nghị Vancouver. Trước năm 2014, hội nghị này được tổ chức tại Long Beach. Các sự kiện của TED cũng được tổ chức ở khắp Bắc Mỹ, châu Âu và châu Á, cung cấp streaming trực tiếp các bài nói chuyện. Họ đề cập đến một loạt các chủ đề trong nghiên cứu và thực hành khoa học và văn hoá, thường thông qua việc kể chuyện. Các diễn giả được cho tối đa là 18 phút để trình bày ý tưởng của mình theo những cách sáng tạo và hấp dẫn nhất mà họ có thể thể hiện. 
- Từ tháng 6 năm 2006, TED Talks đã được cung cấp xem trực tuyến miễn phí, theo điều kiện Ghi công - Phi thương mại - Phi phái sinh của giấy phép Creative Commons, thông qua trang [TED.com](https://www.ted.com/). <img src="https://www.upsieutoc.com/images/2020/08/13/CaptureTed.png" alt="Parrondo" border="0">
- Tính đến tháng 3 năm 2016, hơn 2.400 bài nói chuyện TED Talk đã có sẵn cho phép truy cập tự do trên trang web. Vào tháng 6 năm 2011, con số lượt xem đã gộp vào đã vượt mốc 500 triệu lượt, và vào tháng 11 năm 2012, các bài nói chuyện TED Talk đã được theo dõi trên một tỷ lượt trên toàn thế giới. Tuy vậy, không phải toàn bộ các bài nói chuyện TED Talk đều phổ biến như nhau. Những bài nói chuyện được chia sẻ bởi các học giả có xu hướng được xem trực tuyến nhiều hơn, trong khi các video nghệ thuật và thiết kế có xu hướng được theo dõi ít hơn mức trung bình.

[*Source*](https://vi.wikipedia.org/wiki/TED_(h%E1%BB%99i_th%E1%BA%A3o))

# IMPORT LIBRARY

In [16]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime

# DISCOVERING AND PREPROCESSING DATA

## DISCOVERING DATA

In [22]:
# Đọc file dữ liệu vào
ted = pd.read_csv('ted_main.csv')
# Xem (random rows của dữ liệu) để cảm nhận rõ hơn
ted.sample(10)

comments                                        description  duration  \
490        301  Nina Jablonski says that differing skin colors...       888   
271         66  Robert Lang is a pioneer of the newest kind of...       953   
2273        46  It happens to all of us: you unsubscribe from ...       460   
648        180  Robert Gupta, violinist with the LA Philharmon...       566   
320         87  Rives tells a typographical fairy tale that's ...       197   
2009        42  Basketball is a fast-moving game of improvisat...       728   
1693        64  Vending machines generally offer up sodas, can...       296   
321         41  The Inventables guys, Zach Kaplan and Keith Sc...       946   
1499        62  Young poet, educator and activist Malcolm Lond...       180   
36          67  Ashraf Ghani's passionate and powerful 10-minu...      1125   

                         event   film_date  languages  \
490                    TED2009  1233792000         27   
271                    TED2008  1201910400         29   
2273                 TEDSummit  1467158400         37   
648                    TED2010  1265760000         34   
320                    TED2008  1201910400         49   
2009                   TED2015  1426550400         21   
1693  TED Fellows Retreat 2013  1376870400         32   
321                    TED2005  1107302400         20   
1499       TED Talks Education  1367884800         33   
36              TEDGlobal 2005  1121126400         24   

                     main_speaker  \
490                Nina Jablonski   
271                   Robert Lang   
2273                 James Veitch   
648                  Robert Gupta   
320                         Rives   
2009             Rajiv Maheswaran   
1693          Gabe Barcia-Colombo   
321   Zach Kaplan + Keith Schacht   
1499               Malcolm London   
36                   Ashraf Ghani   

                                                   name  num_speaker  \
490           Nina Jablonski: Skin color is an illusion            1   
271          Robert Lang: The math and magic of origami            1   
2273   James Veitch: The agony of trying to unsubscribe            1   
648    Robert Gupta: Music is medicine, music is sanity            1   
320                   Rives: A story of mixed emoticons            1   
2009  Rajiv Maheswaran: The math behind basketball's...            1   
1693        Gabe Barcia-Colombo: My DNA vending machine            1   
321   Zach Kaplan + Keith Schacht: Toys and material...            2   
1499      Malcolm London: "High School Training Ground"            1   
36          Ashraf Ghani: How to rebuild a broken state            1   

      published_date                                            ratings  \
490       1247792400  [{'id': 22, 'name': 'Fascinating', 'count': 39...   
271       1217422980  [{'id': 1, 'name': 'Beautiful', 'count': 401},...   
2273      1472828989  [{'id': 7, 'name': 'Funny', 'count': 2210}, {'...   
648       1269593640  [{'id': 22, 'name': 'Fascinating', 'count': 23...   
320       1225155600  [{'id': 7, 'name': 'Funny', 'count': 630}, {'i...   
2009      1436195771  [{'id': 22, 'name': 'Fascinating', 'count': 25...   
1693      1394121683  [{'id': 25, 'name': 'OK', 'count': 148}, {'id'...   
321       1225328400  [{'id': 2, 'name': 'Confusing', 'count': 15}, ...   
1499      1368025560  [{'id': 1, 'name': 'Beautiful', 'count': 289},...   
36        1161130260  [{'id': 3, 'name': 'Courageous', 'count': 139}...   

                                          related_talks  \
490   [{'id': 315, 'hero': 'https://pe.tedcdn.com/im...   
271   [{'id': 162, 'hero': 'https://pe.tedcdn.com/im...   
2273  [{'id': 2458, 'hero': 'https://pe.tedcdn.com/i...   
648   [{'id': 218, 'hero': 'https://pe.tedcdn.com/im...   
320   [{'id': 26, 'hero': 'https://pe.tedcdn.com/ima...   
2009  [{'id': 1986, 'hero': 'https://pe.tedcdn.com/i...   
1693  [{'id': 1508, 'hero': 'https://pe.tedcdn.com/i...   
321   [{'

Có hơn 2550 bài nói chuyện. Những điều này đại diện cho tất cả các cuộc đàm phán đã từng được đăng trên Nền tảng TED cho đến ngày 21 tháng 9 năm 2017 và có các cuộc đàm phán được quay trong khoảng thời gian từ năm 1994 đến năm 2017. Đã hơn hai thập kỷ vinh quang của TED.

In [9]:
# Bên cạnh đó, cũng cần phải xem loại dữ liệu được lưu ở kiểu nào
ted.dtypes

comments               int64
description           object
duration               int64
event                 object
film_date              int64
languages              int64
main_speaker          object
name                  object
num_speaker            int64
published_date         int64
ratings               object
related_talks         object
speaker_occupation    object
tags                  object
title                 object
url                   object
views                  int64
dtype: object

In [10]:
# Xem số dòng, số cột
ted.shape

(2550, 17)

In [11]:
# Xem thông tin của dữ liệu
ted.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2550 entries, 0 to 2549
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   comments            2550 non-null   int64 
 1   description         2550 non-null   object
 2   duration            2550 non-null   int64 
 3   event               2550 non-null   object
 4   film_date           2550 non-null   int64 
 5   languages           2550 non-null   int64 
 6   main_speaker        2550 non-null   object
 7   name                2550 non-null   object
 8   num_speaker         2550 non-null   int64 
 9   published_date      2550 non-null   int64 
 10  ratings             2550 non-null   object
 11  related_talks       2550 non-null   object
 12  speaker_occupation  2544 non-null   object
 13  tags                2550 non-null   object
 14  title               2550 non-null   object
 15  url                 2550 non-null   object
 16  views               2550

In [13]:
# Kiểm tra dữ liệu NA
ted.isna().sum()

comments              0
description           0
duration              0
event                 0
film_date             0
languages             0
main_speaker          0
name                  0
num_speaker           0
published_date        0
ratings               0
related_talks         0
speaker_occupation    6
tags                  0
title                 0
url                   0
views                 0
dtype: int64

## PREPROCESSING DATA
- Trong quá trình khám phá dữ liệu, việc quan trọng nhất sau khi tìm hiểu thông tin sơ bộ của dữ liệu thì ta cần phải xử lý các loại dữ liệu bị thiếu. Sau bước `ted.info()` và `ted.isna().sum()` thì ta thấy colum speaker_occupation chỉ có 2544 rows trong khi những colums khác là 2550. Vậy ta cần fill lại như sau:

In [14]:
# Fill missing values
# Có 2 TH fill:
#              1. Missing values là numeric => fill bằng mean
#              2. Missing values là object => fill bằng mode
ted['speaker_occupation'] = ted.speaker_occupation.fillna(ted.speaker_occupation.mode()[0])

In [15]:
# Kiểm tra lại sau khi fill missing values
ted.isna().sum()

comments              0
description           0
duration              0
event                 0
film_date             0
languages             0
main_speaker          0
name                  0
num_speaker           0
published_date        0
ratings               0
related_talks         0
speaker_occupation    0
tags                  0
title                 0
url                   0
views                 0
dtype: int64

- Ngoài ra, sau bước `ted.head()`, có 2 colums `film_date` và `published_date` đang không đúng kiểu hiển thị datetime, nên ta sẽ chuyển sang [kiểu datetime](https://docs.python.org/3/library/datetime.html) tiện xử lý. 

In [17]:
# Chuyển đổi datetime cho film_date và published_date
ted['film_date'] = ted['film_date'].apply(lambda x: datetime.datetime.fromtimestamp(int(x)).strftime('%d-%m-%Y'))
ted['published_date'] = ted['published_date'].apply(lambda x: datetime.datetime.fromtimestamp( int(x)).strftime('%d-%m-%Y'))

In [20]:
# Kiểm tra lại
ted['film_date']
ted['published_date']

0       27-06-2006
1       27-06-2006
2       27-06-2006
3       27-06-2006
4       28-06-2006
           ...    
2545    20-09-2017
2546    20-09-2017
2547    21-09-2017
2548    22-09-2017
2549    22-09-2017
Name: published_date, Length: 2550, dtype: object

## DESCRIBE FEATURES
- `name`: Tên chính thức của ted talk (bao gồm tiêu đề và người nói).
- `title`: Tên của talks
- `description`: Vài nét về nội dung cuộc nói chuyện.
- `main_speaker`: Speakers chính
- `speaker_occupation`: Nghề nghiệp của speakers.
- `num_speaker`: SL speakers.
- `duration`: Thời lượng của talk (tính trên đơn vị giây).
- `event`: Nơi tổ chức TED/TEDx event.
- `film_date`: 
- `published_date`: Ngày up talks lên (Unix timestamps).
- `comments`: SL comments.
- `tags`: Các chủ đề (theme) liên quan đến các talks (vd: science, physics, math, ...).
- `languages`: Số lượng ngôn ngữ được translate của các talks.
- `ratings`: Xếp hạng ratings của các talks.
- `related_talks`: Danh sách từ điển các talks được đề xuất để xem tiếp theo.
- `url`: Link URL gắn với video talk đó.
- `views`: Lượt views của các talks.

# DATA ANALYSIS

## MOST VIEW TALKS
<img src="https://www.upsieutoc.com/images/2020/08/13/top10views.png" alt="Parrondo" border="0">

In [23]:
# Tách dữ liệu để tìm top 10 talks theo 'title', 'main_speaker', 'views', 'film_date'
# sau đó sort theo chiều giảm dần và lấy 10 dòng đầu => top 10
top10_talks = ted[['title', 'main_speaker', 'views', 'film_date']].sort_values('views', ascending=False)[:10]
top10_talks

title       main_speaker  \
0                           Do schools kill creativity?       Ken Robinson   
1346           Your body language may shape who you are          Amy Cuddy   
677                    How great leaders inspire action        Simon Sinek   
837                          The power of vulnerability        Brené Brown   
452              10 things you didn't know about orgasm         Mary Roach   
1776         How to speak so that people want to listen    Julian Treasure   
201                                My stroke of insight  Jill Bolte Taylor   
5                                  Why we do what we do       Tony Robbins   
2114  This is what happens when you reply to spam email       James Veitch   
1416  Looks aren't everything. Believe me, I'm a model.    Cameron Russell   

         views   film_date  
0     47227110  1140825600  
1346  43155405  1340668800  
677   34309432  1253145600  
837   31168150  1275782400  
452   22270883  1233878400  
1776  21594632  1370822400  
201   21190883  1204070400  
5     20685401  1138838400  
2114  20475972  1449532800  
1416  19787465  1351296000